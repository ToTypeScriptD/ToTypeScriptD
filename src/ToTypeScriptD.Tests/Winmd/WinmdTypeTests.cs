using ApprovalTests;
using Mono.Cecil;
using System;
using Xunit;

namespace ToTypeScriptD.Tests.Winmd
{

    public class WinmdTypeTests : WinmdTestBase
    {
        private TypeDefinition GetNativeType(string postfix)
        {
            return base.NativeAssembly.GetNativeType("ToTypeScriptD.Native." + postfix);
        }

        private TypeDefinition GetWinNativeType(string type)
        {
            return base.WindowsAssembly.GetNativeType(type);
        }

        [Fact]
        public void EnumType()
        {
            var result = GetNativeType("SampleEnum").ToTypeScript();
            Approvals.Verify(result);
        }

        [Fact]
        public void EnumTypeWithExplicitValues()
        {
            var result = GetNativeType("SampleEnumNumbered").ToTypeScript();
            Approvals.Verify(result);
        }

        [Fact]
        public void ClassWithEventHandler()
        {
            var result = GetNativeType("ClassWithEventHandler").ToTypeScript();
            Approvals.Verify(result);
        }

        [Fact]
        public void FullSampleAssembly()
        {
            var file = base.NativeAssembly.ComponentPath;
            var errors = new StringBuilderTypeNotFoundErrorHandler();
            var typeCollection = new ToTypeScriptD.Core.TypeWriters.TypeCollection(new ToTypeScriptD.Core.WinMD.WinMDTypeWriterTypeSelector());
            var result = ToTypeScriptD.Render.FullAssembly(file, errors, typeCollection, string.Empty);
            (errors + result).Verify();
        }

        [Fact]
        public void WindowsStorageClass()
        {
            var result = GetWinNativeType("Windows.Storage.StorageFile").ToTypeScript();
            Approvals.Verify(result);
        }
        [Fact]
        public void WindowsSystemUserProfileGlobalizationPreferencesClass()
        {
            var result = GetWinNativeType("Windows.System.UserProfile.GlobalizationPreferences").ToTypeScript();
            Approvals.Verify(result);
        }

        [Fact]
        public void IfEnumNamesAreAllCapsThenTheTranslateToAllLowerCase()
        {
            var result = GetWinNativeType("Windows.Foundation.Metadata.ThreadingModel")
                .ToTypeScript();
            Approvals.Verify(result);
        }

        [Fact]
        public void FullWindowsAssembly()
        {
            var file = @"C:\Windows\System32\WinMetadata\Windows.Foundation.winmd";
            var errors = new StringBuilderTypeNotFoundErrorHandler();
            var typeCollection = new ToTypeScriptD.Core.TypeWriters.TypeCollection(new ToTypeScriptD.Core.WinMD.WinMDTypeWriterTypeSelector());
            var result = ToTypeScriptD.Render.FullAssembly(file, errors, typeCollection, string.Empty);
            (errors + result).Verify();
        }

        //[Fact]
        //public void RenderAllWinmdFilesInWindowsMetadataDirectory()
        //{
        //    var files = @"C:\Windows\System32\WinMetadata\Windows.ApplicationModel.winmd C:\Windows\System32\WinMetadata\Windows.Data.winmd C:\Windows\System32\WinMetadata\Windows.Devices.winmd C:\Windows\System32\WinMetadata\Windows.Foundation.winmd C:\Windows\System32\WinMetadata\Windows.Globalization.winmd C:\Windows\System32\WinMetadata\Windows.Graphics.winmd C:\Windows\System32\WinMetadata\Windows.Management.winmd C:\Windows\System32\WinMetadata\Windows.Media.winmd C:\Windows\System32\WinMetadata\Windows.Networking.winmd C:\Windows\System32\WinMetadata\Windows.Security.winmd C:\Windows\System32\WinMetadata\Windows.Storage.winmd C:\Windows\System32\WinMetadata\Windows.System.winmd C:\Windows\System32\WinMetadata\Windows.UI.winmd C:\Windows\System32\WinMetadata\Windows.UI.Xaml.winmd C:\Windows\System32\WinMetadata\Windows.Web.winmd".Split(' ');
        //    var sw = new System.IO.StringWriter();
        //    var error = new StringBuilderTypeNotFoundErrorHandler();
        //    ToTypeScriptD.Render.AllAssemblies(files, true, sw, error);
        //    var result = error.ToString() + Environment.NewLine + Environment.NewLine + sw.ToString();
        //    result.Verify();
        //}


        [Fact]
        public void AllWinmdFilesInWinMetadata()
        {
            var allFiles = System.IO.Directory.GetFiles(@"C:\Windows\System32\WinMetadata\", "*.winmd");
            var sw = new System.IO.StringWriter();
            var error = new StringBuilderTypeNotFoundErrorHandler();
            var config = new ToTypeScriptD.Core.Config
            {
                OutputType = Core.OutputType.WinRT,
                AssemblyPaths = allFiles,
            };
            ToTypeScriptD.Render.AllAssemblies(config, false, sw, error, string.Empty);
            var result = error.ToString() + Environment.NewLine + Environment.NewLine + sw.ToString();
            result.Verify();
        }

        // Some types in Windows.winmd are duplicated in Windows.Foundation.winmd (huh?)
        //[Fact]
        //public void DumpFullWindowsWinmd()
        //{
        //    var file = @"C:\Program Files (x86)\Windows Kits\8.0\References\CommonConfiguration\Neutral\Windows.winmd";
        //    var result = ToTypeScriptD.Render.FullAssembly(file);
        //    Approvals.Verify(result);
        //}

    }
}
