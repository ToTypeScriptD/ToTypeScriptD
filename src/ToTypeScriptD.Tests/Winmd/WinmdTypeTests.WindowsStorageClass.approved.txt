declare module Windows.Storage {

    export interface IStorageFile extends Windows.Storage.IStorageItem, Windows.Storage.Streams.IRandomAccessStreamReference, Windows.Storage.Streams.IInputStreamReference {
        openAsync(accessMode: Windows.Storage.FileAccessMode): ToTypeScriptD.WinRT.IPromise<Windows.Storage.Streams.IRandomAccessStream>;
        openTransactedWriteAsync(): ToTypeScriptD.WinRT.IPromise<Windows.Storage.StorageStreamTransaction>;
        copyAsync(destinationFolder: Windows.Storage.IStorageFolder): ToTypeScriptD.WinRT.IPromise<Windows.Storage.StorageFile>;
        copyAsync(destinationFolder: Windows.Storage.IStorageFolder, desiredNewName: string): ToTypeScriptD.WinRT.IPromise<Windows.Storage.StorageFile>;
        copyAsync(destinationFolder: Windows.Storage.IStorageFolder, desiredNewName: string, option: Windows.Storage.NameCollisionOption): ToTypeScriptD.WinRT.IPromise<Windows.Storage.StorageFile>;
        copyAndReplaceAsync(fileToReplace: Windows.Storage.IStorageFile): Windows.Foundation.IAsyncAction;
        moveAsync(destinationFolder: Windows.Storage.IStorageFolder): Windows.Foundation.IAsyncAction;
        moveAsync(destinationFolder: Windows.Storage.IStorageFolder, desiredNewName: string): Windows.Foundation.IAsyncAction;
        moveAsync(destinationFolder: Windows.Storage.IStorageFolder, desiredNewName: string, option: Windows.Storage.NameCollisionOption): Windows.Foundation.IAsyncAction;
        moveAndReplaceAsync(fileToReplace: Windows.Storage.IStorageFile): Windows.Foundation.IAsyncAction;
        contentType: string;
        fileType: string;
    }

    export interface IStorageItem {
        renameAsync(desiredName: string): Windows.Foundation.IAsyncAction;
        renameAsync(desiredName: string, option: Windows.Storage.NameCollisionOption): Windows.Foundation.IAsyncAction;
        deleteAsync(): Windows.Foundation.IAsyncAction;
        deleteAsync(option: Windows.Storage.StorageDeleteOption): Windows.Foundation.IAsyncAction;
        getBasicPropertiesAsync(): ToTypeScriptD.WinRT.IPromise<Windows.Storage.FileProperties.BasicProperties>;
        isOfType(type: Windows.Storage.StorageItemTypes): boolean;
        attributes: Windows.Storage.FileAttributes;
        dateCreated: Date;
        name: string;
        path: string;
    }

    export interface IStorageItemProperties {
        getThumbnailAsync(mode: Windows.Storage.FileProperties.ThumbnailMode): ToTypeScriptD.WinRT.IPromise<Windows.Storage.FileProperties.StorageItemThumbnail>;
        getThumbnailAsync(mode: Windows.Storage.FileProperties.ThumbnailMode, requestedSize: number): ToTypeScriptD.WinRT.IPromise<Windows.Storage.FileProperties.StorageItemThumbnail>;
        getThumbnailAsync(mode: Windows.Storage.FileProperties.ThumbnailMode, requestedSize: number, options: Windows.Storage.FileProperties.ThumbnailOptions): ToTypeScriptD.WinRT.IPromise<Windows.Storage.FileProperties.StorageItemThumbnail>;
        displayName: string;
        displayType: string;
        folderRelativeId: string;
        properties: Windows.Storage.FileProperties.StorageItemContentProperties;
    }

    export class StorageFile implements Windows.Storage.IStorageFile, Windows.Storage.IStorageItem, Windows.Storage.Streams.IRandomAccessStreamReference, Windows.Storage.Streams.IInputStreamReference, Windows.Storage.IStorageItemProperties {
        openAsync(accessMode: Windows.Storage.FileAccessMode): ToTypeScriptD.WinRT.IPromise<Windows.Storage.Streams.IRandomAccessStream>;
        openTransactedWriteAsync(): ToTypeScriptD.WinRT.IPromise<Windows.Storage.StorageStreamTransaction>;
        copyAsync(destinationFolder: Windows.Storage.IStorageFolder): ToTypeScriptD.WinRT.IPromise<Windows.Storage.StorageFile>;
        copyAsync(destinationFolder: Windows.Storage.IStorageFolder, desiredNewName: string): ToTypeScriptD.WinRT.IPromise<Windows.Storage.StorageFile>;
        copyAsync(destinationFolder: Windows.Storage.IStorageFolder, desiredNewName: string, option: Windows.Storage.NameCollisionOption): ToTypeScriptD.WinRT.IPromise<Windows.Storage.StorageFile>;
        copyAndReplaceAsync(fileToReplace: Windows.Storage.IStorageFile): Windows.Foundation.IAsyncAction;
        moveAsync(destinationFolder: Windows.Storage.IStorageFolder): Windows.Foundation.IAsyncAction;
        moveAsync(destinationFolder: Windows.Storage.IStorageFolder, desiredNewName: string): Windows.Foundation.IAsyncAction;
        moveAsync(destinationFolder: Windows.Storage.IStorageFolder, desiredNewName: string, option: Windows.Storage.NameCollisionOption): Windows.Foundation.IAsyncAction;
        moveAndReplaceAsync(fileToReplace: Windows.Storage.IStorageFile): Windows.Foundation.IAsyncAction;
        renameAsync(desiredName: string): Windows.Foundation.IAsyncAction;
        renameAsync(desiredName: string, option: Windows.Storage.NameCollisionOption): Windows.Foundation.IAsyncAction;
        deleteAsync(): Windows.Foundation.IAsyncAction;
        deleteAsync(option: Windows.Storage.StorageDeleteOption): Windows.Foundation.IAsyncAction;
        getBasicPropertiesAsync(): ToTypeScriptD.WinRT.IPromise<Windows.Storage.FileProperties.BasicProperties>;
        isOfType(type: Windows.Storage.StorageItemTypes): boolean;
        openReadAsync(): ToTypeScriptD.WinRT.IPromise<Windows.Storage.Streams.IRandomAccessStreamWithContentType>;
        openSequentialReadAsync(): ToTypeScriptD.WinRT.IPromise<Windows.Storage.Streams.IInputStream>;
        getThumbnailAsync(mode: Windows.Storage.FileProperties.ThumbnailMode): ToTypeScriptD.WinRT.IPromise<Windows.Storage.FileProperties.StorageItemThumbnail>;
        getThumbnailAsync(mode: Windows.Storage.FileProperties.ThumbnailMode, requestedSize: number): ToTypeScriptD.WinRT.IPromise<Windows.Storage.FileProperties.StorageItemThumbnail>;
        getThumbnailAsync(mode: Windows.Storage.FileProperties.ThumbnailMode, requestedSize: number, options: Windows.Storage.FileProperties.ThumbnailOptions): ToTypeScriptD.WinRT.IPromise<Windows.Storage.FileProperties.StorageItemThumbnail>;
        static getFileFromPathAsync(path: string): ToTypeScriptD.WinRT.IPromise<Windows.Storage.StorageFile>;
        static getFileFromApplicationUriAsync(uri: Windows.Foundation.Uri): ToTypeScriptD.WinRT.IPromise<Windows.Storage.StorageFile>;
        static createStreamedFileAsync(displayNameWithExtension: string, dataRequested: Windows.Storage.StreamedFileDataRequestedHandler, thumbnail: Windows.Storage.Streams.IRandomAccessStreamReference): ToTypeScriptD.WinRT.IPromise<Windows.Storage.StorageFile>;
        static replaceWithStreamedFileAsync(fileToReplace: Windows.Storage.IStorageFile, dataRequested: Windows.Storage.StreamedFileDataRequestedHandler, thumbnail: Windows.Storage.Streams.IRandomAccessStreamReference): ToTypeScriptD.WinRT.IPromise<Windows.Storage.StorageFile>;
        static createStreamedFileFromUriAsync(displayNameWithExtension: string, uri: Windows.Foundation.Uri, thumbnail: Windows.Storage.Streams.IRandomAccessStreamReference): ToTypeScriptD.WinRT.IPromise<Windows.Storage.StorageFile>;
        static replaceWithStreamedFileFromUriAsync(fileToReplace: Windows.Storage.IStorageFile, uri: Windows.Foundation.Uri, thumbnail: Windows.Storage.Streams.IRandomAccessStreamReference): ToTypeScriptD.WinRT.IPromise<Windows.Storage.StorageFile>;
        contentType: string;
        fileType: string;
        attributes: Windows.Storage.FileAttributes;
        dateCreated: Date;
        name: string;
        path: string;
        displayName: string;
        displayType: string;
        folderRelativeId: string;
        properties: Windows.Storage.FileProperties.StorageItemContentProperties;
    }

}
declare module Windows.Storage.Streams {

    export interface IInputStreamReference {
        openSequentialReadAsync(): ToTypeScriptD.WinRT.IPromise<Windows.Storage.Streams.IInputStream>;
    }

    export interface IRandomAccessStreamReference {
        openReadAsync(): ToTypeScriptD.WinRT.IPromise<Windows.Storage.Streams.IRandomAccessStreamWithContentType>;
    }

}
